Sources/SPMBuildCore/PluginInvocation.swift

```diff
@@ -440,3 +438,32 @@ struct PluginScriptRunnerOutput: Codable {
         let outputFilesDirectory: String
     }
 }
+
+extension ObservabilityMetadata {
+    public var fileLocation: FileLocation? {
+        get {
+            self[FileLocationKey.self]
+        }
+        set {
+            self[FileLocationKey.self] = newValue
+        }
+    }
+
+    private enum FileLocationKey: Key {
+        typealias Value = FileLocation
+    }
+}
+
+public struct FileLocation: Equatable, CustomStringConvertible {
+    let file: String
+    let line: Int?
+
+    public init(_ file: String, line: Int?) {
+        self.file = file
+        self.line = line
+    }
+
+    public var description: String {
+        "\(self.file) \(self.line?.description ?? "")"
```

Comment:
Nitpick: This would end up with a trailing space if there's no line, so maybe `self.line.map{ " \($0.description)" } ?? ""` or however one writes it.